### ISSUE TITLE
Support for banked secondaries for herblore

### ISSUE THREAD

**Issue by hjdarnel** [TheStonedTurtle/banked-experience#9]
Howdy! One improvement I can think of is considering secondaries when calculating banked herblore experience, since your herb xp isn't truly banked unless you have the herbs _and_ the secondaries banked as well. Perhaps some way you could indicate a surplus of secondaries compared to herbs, or vice versa as well?

**Comment by TheStonedTurtle** [comment]
I've thought about implementing this but it becomes difficult for the skills such as Smithing where a single secondary resource (coal) is used for multiple activities. I could ignore cases like this but then the toggle is sorta useless for that skill, i could prioritize higher leveled activities but maybe that isn't what the player wants.

There's also the issue of how to count multiple items as the same secondary/banked item, which is what has prevented me from adding stamina pots to the calculator. Stamina potions would need to combine all the super energy(1)-(4) doses into a single value since its a 1:1 crystal:dose ratio as well as convert marks of grace into crystals. An argument could be made for stuff like super combats where it uses an existing potion that it should also function similarly that since you can decant potions at the GE.

Between deciding on how to handle the secondaries being used in multiple places, counting multiple items as the same secondary, and finding a spot inside the UI i just haven't felt it was worth implementing.

Although it requires some work I just withdraw to my secondary limit so my bank is only calculating what i can make at that moment since items in your inventory are ignored from the calculation.

**Comment by TheStonedTurtle** [comment]
By the way, you can hover over the Secondaries to get a tooltip displaying how much you have.

![https://i.imgur.com/goZHD9D.png](https://i.imgur.com/goZHD9D.png)

If you aren't seeing the secondaries check the plugin config.

**Comment by hjdarnel** [comment]
Gotcha. I understand the complexities of it. I didn’t notice the hover over
the secondaries, that’s pretty helpful for my needs! Thanks. You can close
this if you want, or keep it open if you think you do want to improve it
eventually. Thanks for the plugin :)

**Comment by Budfraiz** [comment]
Be cool if the herblore section could spit out a list of total secondaries needed.

**Comment by GregTheGreek** [comment]
Started chipping away at this, just because I needed a break from work lol.

I think I understand the problem regarding coal, per the screenshot below I think the math for this could work in the same way.
<img width="187" alt="image" src="https://github.com/TheStonedTurtle/banked-experience/assets/16929357/b5507ff3-0753-45ef-9200-a5f9b45a9af4">
Since we can calculate how much is needed per activity (eg: iron, mith, etc) why not offer two sort options: `high` and `low` where high would sort on highest xp earning activity first, and low the inverse. "

I think this is the lowest overhead and shows the true rates for what is banked.

**Comment by TheStonedTurtle** [comment]
Yeah that could work if you make a new dropdown config option where users can select a secondary experience limiting mode and include  `none` for current behavior & default value and `high`/`low` as you recommended.

**Comment by GregTheGreek** [comment]
cool, I'll see how far I get today.

From an architecture POV, would it make the most sense to just keep everything in `SecondaryGrid.java`, I'll create some public getter like `getSecondaryXp()` that would perform a similar calculation to the secondary tooltip?

**Comment by TheStonedTurtle** [comment]
It depends on how you plan to handle all the UI stuff IMO. What visual indicator will there be that an item can't be used because it's missing secondaries (such as mithril ore if not enough coal). Does the calculator need to be able to calculate both all exp and exp limited by secondaries or should it affect the entire calculator based on whether `none` or `low/high` is selected?

But honestly it doesn't really matter, as long as it makes sense where you put it that's fine. Can also make a utility class if you want to, whatever you feel is cleanest.

**Comment by GregTheGreek** [comment]
@TheStonedTurtle I opened up #125, let me know what you think of the first attempt at it

**Comment by griffin1995** [comment]
Hey @TheStonedTurtle,

I read through PR #125 and your review on it, and wanted to have a go at it myself, figured it's worth checking my understanding before I start writing code.

#125 bolted a "Without Secondaries" label next to the main total instead of actually changing how the calculator works. Level projections, progress, all still used the unlimited numbers. `getMissingXp()` had bugs too, skipping secondaries with 0 banked, ignored the user's selected activity in favour of `activities.get(activities.size() - 1)`, and computed total secondary-backed XP rather than the deficit.

I've had a read through the PR and skimmed the codebase, so might be off on some of the technical details, but here's roughly what I'm thinking:

A config dropdown with `None` (default), `High→Low`, `Low→High`. When not `None`, the whole calculator runs in secondary-limited mode. For each secondary, you gather all activities that consume it, sort by the user's selected activity XP in the chosen priority order, then greedily allocate the banked quantity until it runs out. Anything that doesn't get enough is capped.

I think putting the allocation logic in a utility class fed by the existing `secMap`/`availableMap` is probably the way to go, and applying the capping inside `getItemQty()` so totals, level projections, tooltips all pick it up automatically, but I need to look at that in more detail once I get properly into the code.

For the UI, capped items would get a distinct background colour in the grid (same pattern as ignored/RNG), and a collapsible deficit section would show which secondaries are short, by how much, and the XP you're missing.

One thing I noticed (I think) is that `calculateBankedXpTotal()` currently runs before `refreshSecondaries()` builds `secMap`, so I think that order would need to flip when the mode is active, but need to dig into that more. Only runs when the mode is active so `None` is untouched.

I also think it would be a good idea to eventually add a `Custom` mode where you right-click a secondary to manually assign it to specific activities, same engine just user-driven, but thats not a priority rn.

Let me know if this is along the right lines and I'll have a proper look in to it and try to get something built.

Cheers

**Comment by TheStonedTurtle** [comment]
That sounds right but the UI is what I'm most unsure of. Lets say its set to High->Low and you have both `Antifire potion` and `Weapon Poison` select. Both of these take blue dragon scale dust.

You have 100 of each unf potion, `lantadyme` and `kwuarm`, but you only have 150 dust. So all 100 lantadyme will be made (highest level) and then only 50/100 kwuarm will be.

How do we display this? You mentioned:
> For the UI, capped items would get a distinct background colour in the grid (same pattern as ignored/RNG), and a collapsible deficit section would show which secondaries are short, by how much, and the XP you're missing.

I know ignored/RNG already do this but I'd like to avoid adding more color-coded indicators. People with visual impairments may not be able to distinguish between these. Maybe the solution is to store these colors in the config so people can change them?
